### PR TITLE
DEPR offsets: rename 'Q' to 'QE'

### PR DESCRIPTION
**Pull request by natmokval**
xref #9586, #52064

Deprecated the alias denoting quarter end frequency `"Q"` for offsets in favour of `"QE"`.
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v2.2.0.rst`

### CODE REVIEW

**Comment by natmokval** [comment]
@MarcoGorelli, could you please take a look at this PR? It's a follow-up of PR: "DEPR offsets: rename 'M' to 'ME'".
seems like ci failure isn't related to my changes

File: pandas/tests/indexes/datetimes/methods/test_to_period.py [hunk 1]
```diff
@@ -89,6 +89,23 @@ def test_dti_to_period_2monthish(self, freq_offset, freq_period):
...
+            ("2QE-SEP", "2Q-SEP"),
+        ],
+    )
+    def test_to_period_freq_deprecated(self, freq, freq_depr):
+        # GH#9586
+        msg = f"'{freq_depr[1:]}' will be deprecated, please use '{freq[1:]}' instead."
+
+        rng = date_range("01-Jan-2012", periods=8, freq=freq)
+        prng = rng.to_period()
+        with tm.assert_produces_warning(UserWarning, match=msg):
```
Comment (MarcoGorelli) [inline pandas/tests/indexes/datetimes/methods/test_to_period.py:106]: Looks like I'd missed this last time, but this should be FutureWarning, not UserWarning https://github.com/pandas-dev/pandas/pull/52064#pullrequestreview-1690296955

could we address that in a separate PR please?

File: pandas/tests/indexes/period/test_period_range.py [hunk 2]
```diff
@@ -152,3 +160,6 @@ def test_period_range_frequency_ME_error_message(self):
         msg = "Invalid frequency: 2ME"
         with pytest.raises(ValueError, match=msg):
             period_range(start="Jan-2000", end="Dec-2000", freq="2ME")
+
+    def test_my(self):
+        Period("2017Q1", freq="Q")
```
Comment (MarcoGorelli) [inline pandas/tests/indexes/period/test_period_range.py:165]: temporary test name?😉

File: pandas/tests/scalar/period/test_period.py [hunk 3]
```diff
@@ -73,10 +73,8 @@ def test_construction(self):
 
         i1 = Period.now(freq="Q")
         i2 = Period(datetime.now(), freq="Q")
-        i3 = Period.now("q")
```
Comment (MarcoGorelli) [inline pandas/tests/scalar/period/test_period.py:76]: could you remind me why this is removed please?

File: pandas/tests/indexes/datetimes/methods/test_to_period.py (hunk 1 above)
Comment (natmokval) [reply pandas/tests/indexes/datetimes/methods/test_to_period.py:106]: thank you, I corrected Warning class and made the separate PR

File: doc/source/whatsnew/v2.2.0.rst [hunk 4]
```diff
@@ -232,6 +232,28 @@ For example:
 
     pd.date_range('2020-01-01', periods=3, freq='ME')
 
+Deprecate alias ``Q`` in favour of ``QE`` for offsets
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.2.0.rst:236]: This looks very similar to the previous section, perhaps we can combine them?

File: pandas/io/json/_table_schema.py [hunk 5]
```diff
@@ -207,6 +208,8 @@ def convert_json_field_to_pandas_type(field) -> str | CategoricalDtype:
         if field.get("tz"):
             return f"datetime64[ns, {field['tz']}]"
         elif field.get("freq"):
+            # GH#9586 rename frequency M to ME for offsets
+            field["freq"] = freq_to_period_freqstr(1, field["freq"])
```
Comment (MarcoGorelli) [inline pandas/io/json/_table_schema.py:212]: left over from https://github.com/pandas-dev/pandas/pull/55581#pullrequestreview-1693046244? should we wait for that one to be resolved first?

**Review by MarcoGorelli — COMMENTED** [review]
looks good, just this part isn't totally clear to me https://github.com/pandas-dev/pandas/pull/55553/files#r1367138011

File: doc/source/whatsnew/v2.2.0.rst (hunk 4 above)
Comment (natmokval) [reply doc/source/whatsnew/v2.2.0.rst]: yes, I agree. I combined these two sections together.

File: pandas/io/json/_table_schema.py (hunk 5 above)
Comment (natmokval) [reply pandas/io/json/_table_schema.py]: yes, it' related to #55581. I agree, it's better at first to merge #55581. I rolled my changes in `pandas/io/json/_table_schema.py` back , but now I have failures in `test_read_json_table_period_orient`, which will be resolved after merging.

File: pandas/tests/scalar/period/test_period.py (hunk 3 above)
Comment (MarcoGorelli) [reply pandas/tests/scalar/period/test_period.py:76]: as is it just because we're not uppercasing qe? if so, sounds fine tbh

**Review by MarcoGorelli — APPROVED** [review]
cool, should be good on green!

File: pandas/tests/scalar/period/test_period.py (hunk 3 above)
Comment (natmokval) [reply pandas/tests/scalar/period/test_period.py:76]: we had a strange behaviour, the same ass for deprecation M to ME:

```
Period('2020-01', 'QE') => errors
Period('2020-01', 'qe') => silently converts to 'M'
Period('2020-01', 'q') => errors
```
I added `'qe' `to `_dont_uppercase` and it fixed the problem. I left a comment linking to #54105, saying that this should raise in the future an error.

**Comment by MarcoGorelli** [comment]
looks like there's some test failures
```
 FAILED pandas/tests/io/json/test_json_table_schema.py::TestTableOrientReader::test_read_json_table_period_orient[vals0-None] - TypeError: data type 'period[QE-DEC]' not understood
```

**Comment by natmokval** [comment]
> looks like there's some test failures
> 
> ```
>  FAILED pandas/tests/io/json/test_json_table_schema.py::TestTableOrientReader::test_read_json_table_period_orient[vals0-None] - TypeError: data type 'period[QE-DEC]' not understood
> ```

`test_read_json_table_period_orient` failed because there was a bug for conversion JSON field descriptor into pandas type for deprecated offsets frequency 'M'. We fixed it in #55581, now the test passes.

I added parameters `'Q'` and `'2Q'` into parametrization of `test_read_json_table_orient_period_depr_freq`

**Comment by MarcoGorelli** [comment]
thanks @natmokval !

**Pull request merged**